reference-implementation/lib/writable-stream.js

```diff
@@ -7,222 +7,462 @@ const { DequeueValue, EnqueueValueWithSize, GetTotalQueueSize, PeekQueueValue }
 const CountQueuingStrategy = require('./count-queuing-strategy.js');
 
 class WritableStream {
-  constructor(underlyingSink = {}, { size, highWaterMark = 0 } = {}) {
-    this._underlyingSink = underlyingSink;
+  constructor(underlyingSink = {}, { size, highWaterMark } = {}) {
+    this._state = 'writable';
 
-    this._closedPromise = new Promise((resolve, reject) => {
-      this._closedPromise_resolve = resolve;
-      this._closedPromise_reject = reject;
-    });
+    this._writer = undefined;
+    this._storedError = undefined;
 
-    this._readyPromise = Promise.resolve(undefined);
-    this._readyPromise_resolve = null;
+    // Initialize to undefined first because the constructor of the controller checks this
+    // variable to validate the caller.
+    this._writableStreamController = undefined;
+    const type = underlyingSink.type;
+    if (type === undefined) {
+      if (highWaterMark === undefined) {
+        highWaterMark = 0;
+      }
+      this._writableStreamController = new WritableStreamDefaultController(this, underlyingSink, size, highWaterMark);
+    } else {
+      throw new RangeError('Invalid type is specified');
+    }
+  }
 
-    this._queue = [];
-    this._state = 'writable';
-    this._started = false;
-    this._writing = false;
+  get locked() {
+    if (IsWritableStream(this) === false) {
+      throw new TypeError('WritableStream.prototype.locked can only be used on a WritableStream');
+    }
 
-    const normalizedStrategy = ValidateAndNormalizeQueuingStrategy(size, highWaterMark);
-    this._strategySize = normalizedStrategy.size;
-    this._strategyHWM = normalizedStrategy.highWaterMark;
+    return IsWritableStreamLocked(this);
+  }
+
+  getWriter() {
+    if (IsWritableStream(this) === false) {
+      throw new TypeError('WritableStream.prototype.getWriter can only be used on a WritableStream');
+    }
+
+    return AcquireWritableStreamDefaultWriter(this);
+  }
```

Comment:
Although I know there is a thread on abort(), for now (not having looked at that thread in detail yet) I think we should keep abort() as a lock-abort-unlock, similar to how cancel() on readable streams is lock-cancel-unlock.